[AIRFLOW-5549] Extended BQ GetDataOperator to handle query params

mohannadbanayosi opened the pull request:
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-5549) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AIRFLOW-5549

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Currently the BigQueryGetDataOperator returns a list of rows only and not the total number of rows available and the page token to navigate to the next page. It also does not handle pagination (to request the next set of rows). The operator will now return the total number of rows and the next page token as well as handling start_index and page_token parameters.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

feluelle commented on airflow/gcp/operators/bigquery.py:
```diff
@@ -217,7 +217,10 @@ class BigQueryGetDataOperator(BaseOperator):
-    **Example Result**: ``[['Tony', '10'], ['Mike', '20'], ['Steve', '15']]``
+    .. seealso:: https://cloud.google.com/bigquery/docs/reference/v2/tabledata/list
+
+    **Example Result**: ``{'totalRows': 20, 'pageToken': 'BHSHJT3DRAFSSIAICAFCACSB77777777757SUAA=',
```
You can use `.. code-block:: json` for better formatting

feluelle commented on airflow/gcp/operators/bigquery.py:
```diff
@@ -259,6 +267,9 @@ class BigQueryGetDataOperator(BaseOperator):
     :type location: str
+
+    :rtype: dict
+        Map containing the requested rows + metadata.
```
That's the return value of the `execute` function, isn't it? `:rtype: dict` will show up in an IDE that `__init__` is returning a dict which is wrong. You also find the return value of `__init__` in line 290 (`-> None:`)

feluelle commented on airflow/gcp/operators/bigquery.py:
```diff
@@ -242,9 +245,14 @@ class BigQueryGetDataOperator(BaseOperator):
     :param max_results: The maximum number of records (rows) to be fetched
         from the table. (templated)
-    :type max_results: str
+    :type max_results: int
```
If we change the type of argument we need to mention that in the UPDATING.md so that people will see that before they run into an error.

mohannadbanayosi replied:
Didn't know that, thanks for clarifying. Will update it now.

mik-laj replied:
This is only required if the change is not backward compatible. In this case, is backward compatible.

feluelle commented on airflow/gcp/operators/bigquery.py:
```diff
@@ -285,37 +299,45 @@ def __init__(self,
         rows = response['rows']
 
-        table_data = []
+        table_rows = []
```
You could also just do:
```suggestion
        table_rows = [
            [fields['v'] for fields in dict_row['f']]
            for dict_row in rows
        ]
```
(my preference)

mohannadbanayosi replied:
I prefer this too, but didn't want to change it for readability.

feluelle commented on airflow/gcp/operators/bigquery.py:
```diff
@@ -217,7 +217,17 @@ class BigQueryGetDataOperator(BaseOperator):
-    **Example Result**: ``[['Tony', '10'], ['Mike', '20'], ['Steve', '15']]``
+    .. seealso:: https://cloud.google.com/bigquery/docs/reference/v2/tabledata/list
+
+    **Example Result**: ::
```
```suggestion
    **Example Result**:
```
I am not sure if you need those colons. You can check by building the docs via `./breeze --build-docs` if it makes a difference.

mohannadbanayosi replied:
It fails if I remove the `::`

feluelle commented on tests/gcp/hooks/test_bigquery.py:
```diff
@@ -460,6 +460,49 @@ def test_insert_all_fail(self):
+        mock_service = mock.Mock()
+        with mock.patch.object(hook.BigQueryBaseCursor(mock_service, project_id).service,
+                               'tabledata') as MockService:
+            MockService.return_value.list(
```
Do you really want to call the mock `(...)`? Or do you need just the return value (`.return_value`)?
```suggestion
            MockService.return_value.list.return_value.execute.return_value = expected_result
```
In your current test you won't see the difference because you don't check for calls of mocks.
I am always in favor of checking the mock calls of the objects you use for example via `assert_called_once_with()`.
See the [docs](https://docs.python.org/3/library/unittest.mock.html) for more information.

feluelle commented on tests/gcp/hooks/test_bigquery.py:
```diff
@@ -460,6 +460,49 @@ def test_insert_all_fail(self):
+        table_id = 'table'
+
+        mock_service = mock.Mock()
+        with mock.patch.object(hook.BigQueryBaseCursor(mock_service, project_id).service,
```
I don't know if you know, but you can also use the `@mock.patch.object` decorator/annotation. I like it more especially when basically your whole test is using the mock object.

I think it also reduces the _clutter_ in tests.

feluelle commented on tests/gcp/hooks/test_bigquery.py:
```diff
@@ -460,6 +460,49 @@ def test_insert_all_fail(self):
+            result = hook.BigQueryBaseCursor(
+                mock_service, project_id).get_tabledata(
+                dataset_id=dataset_id, table_id=table_id)
+            self.assertEqual(result, expected_result)
```
I try to split up my test into three sections.

1. Setup (mostly via annotations/decorator)
2. The actual call
3. The assertion

and put one space in between each section like so https://github.com/apache/airflow/blob/master/tests/contrib/hooks/test_imap_hook.py so that you can directly see what the test is trying to cover.

feluelle reviewed (commented):
I just noticed that the return type was wrong before. The table data was a list not a dict.

Could you mention that in the UPDATING.md, too? That the return type changed from list to dict.

mohannadbanayosi committed:
[AIRFLOW-5549] Extended BQ GetDataOperator to handle query params

The operator will now return the total number of rows and the next page token as well as handling start_index and page_token parameters.

Signed-off-by: Mohannad Albanayosy <m.banayosi@gmail.com>

mik-laj commented on airflow/gcp/operators/bigquery.py:
```diff
@@ -242,9 +252,14 @@ class BigQueryGetDataOperator(BaseOperator):
         from the table. (templated)
-    :type max_results: str
+    :type max_results: int
+    :param page_token: (Optional) Page token, returned from a previous call,
```
This is the implementation detail of the library. This should only be in the hook, and the hook method should always return the full range of data. Otherwise, It will not possible to migrate to different library eg. https://pypi.org/project/google-cloud-bigquery/ 

My team is working on migrating integration(if possible) on google-cloud-python. It's a rpart of recommendation of GCP integration guide. We have started retrofiting this operator also - https://github.com/PolideaInternal/airflow/issues/231

@TobKed Can you provide a detail information about retrofiting integration with BigQuery?

TobKed replied:
IMO page token should should not be passed as a parameter to Operator, maybe in Hook but also I don't think it is good idea, as @mik-laj said: 

> the hook method should always return the full range of data

A method in `google-cloud-bigquery` accepts `page_token` parameter hence it may be optionally used in hook
https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.client.Client.html?highlight=list_rows#google.cloud.bigquery.client.Client.list_rows

WDYT @mik-laj @nuclearpinguin

mik-laj commented on UPDATING.md:
```diff
@@ -40,6 +40,29 @@ assists users migrating to a new version.
 
+### Changes to BigQueryGetDataOperator
+`max_results` parameter type changed from `str` to `int` to reflect the correct data type 
+and `location` parameter removed as it is redundant.
```
@nuclearpinguin Have you added this parameter recently? Can you give more explanations?

TobKed replied:
PR created by @nuclearpinguin https://github.com/apache/airflow/pull/6020/files.
IMO this change seems to be against this PR and related JIRA's

turbaszek replied:
The location in used in the `get_tabledata` method which is used in execute.
https://github.com/apache/airflow/blob/65851f7a747a61ac486f18ec9b5dfcccb711027e/airflow/gcp/hooks/bigquery.py#L1489-L1508

TobKed commented on airflow/gcp/operators/bigquery.py:
```diff
@@ -217,7 +217,17 @@ class BigQueryGetDataOperator(BaseOperator):
     be equal to the number of rows fetched. Each element in the list will again be a list
     where element would represent the columns values for that row.
 
```
This description seems to be not valid anymore. If someone is checking the number of the elements of returned list (and it is dictionary) it will be always three.
